bin/l10n-sync.js

```diff
@@ -0,0 +1,352 @@
+ * of how many times that occurs.
+ *
+ * It's a pretty simple hack and would be good to have a more sophisticated
+ * (localized?) API function. But it's not really worth for a deployment only
+ * script.
+ */
+function fewTimes(count /*: number */) /*: string */ {
+  switch (count) {
+    case 1:
+      return 'once';
+    case 2:
+      return 'twice';
+    default:
+      return `${count} times`;
+  }
+}
+
+/**
+ * Tries to sync the l10n branch and retries for 3 times if it fails to sync.
+ *
+ * @throws Will throw an error if it fails to sync for more than 3 times.
+ */
+async function tryToSync(upstream /*: string */) /*: Promise<void> */ {
+  console.log('>>> Syncing the l10n branch with main.');
+  // RegExp for matching only the vendored locales.
+  // It matches the files in `locale` directory but excludes `en-US` which is the
+  // main locale we edit.
+  const vendoredLocalesPath = /^locales[\\/](?!en-US[\\/])/;
+  // RegExp for matching anything but the vendored locales.
+  // It matches the files that are not in the `locales` directory EXCEPT the
+  // `locales/en-US/` directory which is the main locale we edit.
+  const nonVendoredLocalesPath = /^(locales[\\/]en-US[\\/]|(?!locales[\\/]))/;
+  // We have a total try count to re-try to sync a few more time if the previous
+  // try fails. This can accur when someone else commits to l10n branch while we
+  // are syncing. In that case, `git push` will fail and we'll pull the latest
+  // changes and try again. Nevertheless, we should have a hard cap on the try
+  // count for safety.
+  const totalTryCount = 3;
+  let error /*: Error | null */ = null;
+  let tryCount = 0;
+
+  // Try to sync and retry for `totalTryCount` times if it fails.
+  do {
+    try {
+      if (tryCount > 0) {
+        console.warn(stripIndent`
+        Syncing the l10n branch has failed.
+        This may be due to a new commit during this operation. Trying again.
+        Tried ${fewTimes(tryCount)} out of ${totalTryCount}.
+      `);
+      }
+
+      console.log(`>>> Fetching upstream ${upstream}.`);
+      await logAndExec('git', 'fetch', upstream);
+
+      // First, check if the l10n branch contains only changes in `locales` directory.
+      await checkAllowedPaths({
+        upstream,
+        compareBranch: 'l10n',
+        baseBranch: 'main',
+        matchRegexp: vendoredLocalesPath,
+      });
+
+      // Second, check if the main branch contains changes except the translated locales.
+      await checkAllowedPaths({
+        upstream,
+        compareBranch: 'main',
+        baseBranch: 'l10n',
+        matchRegexp: nonVendoredLocalesPath,
+      });
+
+      console.log('>>> Merging main to l10n.');
+      await logAndExec('git', 'checkout', `${upstream}/l10n`);
+
+      const currentDate = DATE_FORMAT.format(new Date());
+      await logAndExec(
+        'git',
+        'merge',
+        `${upstream}/main`,
+        '-m',
+        `🔃 Daily sync: main -> l10n (${currentDate})`
```

Comment:
I wonder if the message shouldn't be extracted to a constant at the start of the file (just the constant part), what do you think?

Reply:
Yeah, it would be good to extract that part. Just did it.